replica/distributed_loader.cc

```diff
@@ -490,6 +492,35 @@ future<> distributed_loader::populate_keyspace(distributed<replica::database>& d
             });
         }
     });
+
+    // here we can be sure that the 'truncated' table is loaded,
+    // now we load and initialize the truncation times for the tables
+    {
+        const auto truncation_times = co_await sys_ks.local().load_truncation_times();
+        co_await smp::invoke_on_all([&] {
+            db.local().get_tables_metadata().for_each_table([&](table_id id, lw_shared_ptr<table> table) {
+                if (table->schema()->ks_name() != ks_name) {
+                    return;
+                }
+                const auto it = truncation_times.find(id);
+                const auto truncation_time = it == truncation_times.end()
+                    ? db_clock::time_point::min()
+                    : it->second;
+                if (this_shard_id() == 0 && table->schema()->ks_name() == db::schema_tables::NAME &&
+                    truncation_time != db_clock::time_point::min()) {
+                    // replay_position stored in the truncation record may belong to
+                    // the old (default) commitlog domain. It's not safe to interpret
+                    // that replay position in the schema commitlog domain.
+                    // Refuse to boot in this case. We assume no one truncated schema tables.
+                    // We will hit this during rolling upgrade, in which case the user will
+                    // roll back and let us know.
+                    throw std::runtime_error(format("Schema table {}.{} has a truncation record. Booting is not safe.",
+                        table->schema()->ks_name(), table->schema()->cf_name()));
```

Comment:
Not a comment on the patch, but how does one recover from this situation?

I guess one could move the truncation table sstables to another cluster, remove the record, and move the sstables back.